### PR TITLE
wal: fix the unexpectedEOF error in the last wal.

### PR DESCRIPTION
**Pull request by xiang90**
It is safe to repair the unexpectedEOF error in the last wal. raft
will not send out message before the entry successfully committed
into wal. Thus we can safely truncate the last entry in the wal
to repair.

### CODE REVIEW

File: etcdserver/storage.go [hunk 1]
```diff
@@ -72,13 +73,28 @@ func (st *storage) SaveSnap(snap raftpb.Snapshot) error {
...
-	}
-	var wmetadata []byte
-	if wmetadata, st, ents, err = w.ReadAll(); err != nil {
-		log.Fatalf("etcdserver: read wal error: %v", err)
+	var (
+		err       error
+		wmetadata []byte
+	)
+
+	for i := 0; i < 2; i++ {
```
Comment (yichengq) [inline etcdserver/storage.go]: it is hard to understand why it uses a for loop of 2 here.
What about always do the integrity check as the first step?

File: wal/wal.go [hunk 2]
```diff
@@ -414,14 +414,14 @@ func (w *WAL) Save(st raftpb.HardState, ents []raftpb.Entry) error {
...
-		return err
-	}
 	for i := range ents {
 		if err := w.saveEntry(&ents[i]); err != nil {
 			return err
 		}
 	}
+	if err := w.saveState(&st); err != nil {
+		return err
+	}
```
Comment (yichengq) [inline wal/wal.go]: why do we switch the order of saves?

File: etcdserver/storage.go (hunk 1 above)
Comment (xiang90) [inline etcdserver/storage.go]: i do not like that idea for 2 is a usual pattern for fall back and repair

File: wal/wal.go (hunk 2 above)
Comment (xiang90) [inline wal/wal.go]: since the previous order one is incorrect. commit index should be saved after the entry is saved

File: etcdserver/storage.go (hunk 1 above)
Comment (yichengq) [inline etcdserver/storage.go]: :ok:

File: wal/wal.go (hunk 2 above)
Comment (yichengq) [inline wal/wal.go]: Could we split it into another commit?

File: etcdserver/storage.go [hunk 3]
```diff
@@ -72,13 +73,28 @@ func (st *storage) SaveSnap(snap raftpb.Snapshot) error {
...
+		if w, err = wal.Open(waldir, snap); err != nil {
+			log.Fatalf("etcdserver: open wal error: %v", err)
+		}
+		if wmetadata, st, ents, err = w.ReadAll(); err != nil {
+			w.Close()
+			if i != 0 || err != io.ErrUnexpectedEOF {
+				log.Fatalf("etcdserver: read wal error: %v", err)
+			}
+			if !wal.Repair(waldir) {
+				log.Fatalf("etcdserver: %v cannot be fixed", err)
```
Comment (yichengq) [inline etcdserver/storage.go]: %v in WAL

File: wal/repair.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	names = checkWalNames(names)
+	if len(names) == 0 {
+		return nil, ErrFileNotFound
+	}
+	last := path.Join(dirpath, names[len(names)-1])
+	f, err := os.OpenFile(last, os.O_RDWR, 0)
+	if err != nil {
+		return nil, err
+	}
+	return f, nil
```
Comment (yichengq) [inline wal/repair.go]: return os.OpenFile(last, os.O_RDWR, 0)

File: wal/repair.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		switch err {
+		case nil:
+			n += 8 + rec.Size()
+			continue
+		case io.EOF:
+			return true
+		case io.ErrUnexpectedEOF:
+			err = f.Truncate(int64(n))
+			if err != nil {
+				return false
```
Comment (yichengq) [inline wal/repair.go]: log the error?

File: wal/wal.go (hunk 2 above)
Comment (xiang90) [inline wal/wal.go]: maybe.

File: wal/repair.go (hunk 4 above)
Comment (xiang90) [inline wal/repair.go]: ok

File: wal/repair.go (hunk 5 above)
Comment (xiang90) [inline wal/repair.go]: this is logged in etcdserver. i do not think it is necessary to log the specific error(user cannot solve it anyway) and it is the reason this function return a bool rather than error.

File: etcdserver/storage.go (hunk 1 above)
Comment (philips) [inline etcdserver/storage.go]: this does deserve a comment; I have no idea why 2.

File: etcdserver/storage.go [hunk 6]
```diff
@@ -72,13 +73,31 @@ func (st *storage) SaveSnap(snap raftpb.Snapshot) error {
...
+
+	repaired := false
+	for {
+		if w, err = wal.Open(waldir, snap); err != nil {
+			log.Fatalf("etcdserver: open wal error: %v", err)
+		}
+		if wmetadata, st, ents, err = w.ReadAll(); err != nil {
+			w.Close()
+			// we can only repair ErrUnexpectedEOF and we never repair twice.
+			if repaired || err != io.ErrUnexpectedEOF {
```
Comment (xiang90) [inline etcdserver/storage.go]: @philips OK. How about this?

File: wal/repair.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+			if err = f.Truncate(int64(n)); err != nil {
+				log.Printf("wal: could not repair %v, failed to truncate file", f.Name())
+				return false
+			}
+			if err = f.Sync(); err != nil {
+				log.Printf("wal: could not repair %v, failed to sync file", f.Name())
+				return false
+			}
+			return true
+		}
```
Comment (yichengq) [inline wal/repair.go]: please handle the other error cases.

In practice for current version, it may go out with error 'crc mismatch'(this should be fixed by update crc for decoder), and truncate at the wrong place.

**Comment by xiang90** [comment]
@yichengq fixed. please try again. :)

File: wal/repair.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+				log.Printf("wal: could not repair %v, failed to truncate file", f.Name())
+				return false
+			}
+			if err = f.Sync(); err != nil {
+				log.Printf("wal: could not repair %v, failed to sync file", f.Name())
+				return false
+			}
+			return true
+		default:
+			log.Printf("wal could not repair error (%v)", err)
```
Comment (yichengq) [inline wal/repair.go]: wal:

File: wal/util.go [hunk 9]
```diff
@@ -76,8 +82,11 @@ func checkWalNames(names []string) []string {
 }
 
 func parseWalName(str string) (seq, index uint64, err error) {
+	if !strings.HasSuffix(str, ".wal") {
+		return 0, 0, badWalName
+	}
 	_, err = fmt.Sscanf(str, "%016x-%016x.wal", &seq, &index)
-	return
+	return seq, index, err
```
Comment (yichengq) [inline wal/util.go]: testing for this parse?

Comment (xiang90) [inline wal/util.go]: i do not think we have a test for parsing. i do not think it is a blocker for landing this pr, which fixes the error issue.

**Comment by yichengq** [comment]
lgtm
